File: 5.py
import sys
import itertools

class TSP:
    def __init__(self, num_cities, distances):
        self.num_cities = num_cities
        self.distances = distances
        self.visited = [False] * num_cities
        self.min_cost = sys.maxsize
        self.min_path = []

    def branch_and_bound(self):
        self._branch_and_bound(0, 0, [0])

    def _branch_and_bound(self, current_city, current_cost, current_path):
        if len(current_path) == self.num_cities:
            current_cost += self.distances[current_path[-1]][current_path[0]]
            if current_cost < self.min_cost:
                self.min_cost = current_cost
                self.min_path = current_path.copy()
            return

        for next_city in range(self.num_cities):
            if not self.visited[next_city]:
                bound = self.calculate_bound(current_path, current_cost, next_city)
                if bound < self.min_cost:
                    self.visited[next_city] = True
                    current_path.append(next_city)
                    self._branch_and_bound(next_city, current_cost + self.distances[current_city][next_city], current_path)
                    current_path.pop()
                    self.visited[next_city] = False

    def calculate_bound(self, current_path, current_cost, next_city):
        bound = current_cost

        # Lower bound: adding the minimum edge cost from the current city
        bound += min([self.distances[current_city][next_city] for current_city in range(self.num_cities) if current_city not in current_path])

        # Upper bound: adding the minimum edge cost to complete the tour
        bound += min([self.distances[next_city][city] for city in range(self.num_cities) if city not in current_path])

        return bound

def main():
    print("Traveling Salesman Problem Solver (Branch and Bound)")
    num_cities = int(input("Enter the number of cities: "))

    distances = []
    print("Enter the distance matrix (enter -1 for infinity):")
    for i in range(num_cities):
        row = [int(x) for x in input().split()]
        distances.append(row)

    tsp_solver = TSP(num_cities, distances)
    tsp_solver.branch_and_bound()

    print("\nOptimal Path:", tsp_solver.min_path)
    print("Total Cost:", tsp_solver.min_cost)

if __name__ == "__main__":
    main()

File: 5.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

void displayMatrix(const vector<vector<int>>& graph) {
    int V = graph.size();
    cout << "Distance Matrix:\n";
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            cout << graph[i][j] << " ";
        }
        cout << endl;
    }
}

int travllingSalesmanProblem(vector<vector<int>>& graph, int s) {
    int V = graph.size();
    vector<int> vertex;
    
    for (int i = 0; i < V; i++) {
        if (i != s) {
            vertex.push_back(i);
        }
    }
    int min_path = INT_MAX;
    do {
        int current_pathweight = 0;
        int k = s;
        for (int i = 0; i < vertex.size(); i++) {
            current_pathweight += graph[k][vertex[i]];
            k = vertex[i];
        }
        current_pathweight += graph[k][s];
        min_path = min(min_path, current_pathweight);
    } while (next_permutation(vertex.begin(), vertex.end()));

    return min_path;
}

int main() {
    int V;
    cout << "Enter the number of cities: ";
    cin >> V;
    vector<vector<int>> graph(V, vector<int>(V, 0));
    cout << "Enter distances from one city to another (enter -1 for unreachable cities):\n";
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            if (i != j) {
                cout << "Distance from City " << i << " to City " << j << ": ";
                cin >> graph[i][j];
            } else {
                graph[i][j] = 0;  // Distance from a city to itself is 0.
            }
        }
    }
    displayMatrix(graph); // Display the distance matrix
    int s;
    cout << "Enter the starting city (0 to " << V - 1 << "): ";
    cin >> s;
    if (s < 0 || s >= V) {
        cout << "Invalid starting city.\n";
        return 1;
    }
    int minPath = travllingSalesmanProblem(graph, s);
    cout << "Minimum path cost: " << minPath << endl;
    return 0;
}

